tools/azure-sdk-tools/ci_tools/dependency_analysis.py

```diff
@@ -0,0 +1,350 @@
+    template = env.get_template("deps.html.j2")
+    with io.open(output_path, "w", encoding="utf-8") as output:
+        output.write(template.render(report_context))
+
+
+def get_dependent_packages(data_pkgs) -> Set[str]:
+    # Get unique set of Azure SDK packages that are added as required package
+    deps = []
+    for v in data_pkgs.values():
+        deps.extend([dep["name"] for dep in v["deps"] if not dump_should_skip_lib(dep["name"])])
+    return set(deps)
+
+
+def dump_packages(data_pkgs: Dict[str, Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
+    dump_data = {}
+    unique_dependent_packages = get_dependent_packages(data_pkgs)
+    for p_name, p_data in data_pkgs.items():
+        p_id = p_name + ":" + p_data["version"]
+        dep = [p for p in p_data["deps"] if not dump_should_skip_lib(p["name"])]
+        # Add package if it requires other azure sdk package or if it is added as required by other sdk package
+        if len(dep) > 0 or p_name in unique_dependent_packages:
+            dump_data[p_id] = {"name": p_name, "version": p_data["version"], "type": "internal", "deps": dep}
+    return dump_data
+
+
+def resolve_lib_deps(dump_data: Dict[str, Dict[str, Any]], data_pkgs: Dict[str, Dict[str, Any]], pkg_id: str) -> None:
+    for dep in dump_data[pkg_id]["deps"]:
+        dep_req = Requirement.parse(dep["name"] + dep["version"])
+        if dep["name"] in data_pkgs and data_pkgs[dep["name"]]["version"] in dep_req:
+            # If the internal package version matches the dependency spec,
+            # rewrite the dep version to match the internal package version
+            dep["version"] = data_pkgs[dep["name"]]["version"]
+        else:
+            dep_id = dep["name"] + ":" + dep["version"]
+            if not dep_id in dump_data:
+                dump_data[dep_id] = {
+                    "name": dep["name"],
+                    "version": dep["version"],
+                    "type": "internalbinary" if dep["name"] in data_pkgs else "external",
+                    "deps": [],
+                }
+
+
+def pluralize(data: any, singular: str, plural: str) -> str:
+    if len(data) == 1 or data is None:
+        return singular
+    else:
+        return plural
+
+
+def analyze_dependencies() -> None:
+    base_dir = discover_repo_root()
+
+    parser = argparse.ArgumentParser(
+        description="""\
+    Analyze dependencies in Python packages. First, all declared dependencies
+    and the libraries that declare them will be discovered (visible with
+    --verbose). Next, all declared dependency version specs will be analyzed to
+    ensure that none of requirements conflict with any other package in the repo. Finally, all declared dependency
+    version specs will be frozen to shared_requirements.txt.
+    """
+    )
+    parser.add_argument("--verbose", help="verbose output", action="store_true")
+    parser.add_argument(
+        "--freeze",
```

Comment:
When should this be used?

Reply:
Down below on line 301.

if `args.freeze` is true, we write the requirements to `shared_requirements.txt`

Reply:
This is partially why I didn't make the lack of `shared_requirements.txt` an exitcode=1 scenario to start with.

> if they don't _have_ a shared_requirements.txt to begin with should that always be an error exit?

Reply:
Ah I think I assumed that this script would only run in CI. Just so I'm super clear, if I need to update the shared_requirements.txt for my PR, I should run this script with the `--freeze` arg?

Reply:
Exactly! And if your new PR _doesn't_ add any new shared_requirements, it'll just pass 👍 

So new external deps will need to be approved by Anna/Laurent/SomeoneElseOnPythonTeamLikeJohanMaybe, but that's it!

No more `override`. That entire concept is ripped out.

Reply:
Gotcha! Okay that clears up everything for me. 😃